Controller/ContentManagement/CrudController.php

```diff
@@ -358,4 +374,47 @@ public function getContentTypeInfo(ContentType $contentType)
                 'contentType' => $contentType,
         ]);
     }
+
+    /**
+     * @Route("/api/user-profile", defaults={"_format": "json"}, methods={"GET"})
+     */
+    public function getUserProfile() : JsonResponse
+    {
+        $user = $this->getUser();
+        if (! $user instanceof User) {
+            throw new \RuntimeException('User profile class not recognized');
+        }
+        if (! $user->isEnabled()) {
+            throw new \RuntimeException('User disabled');
+        }
+
+        return $this->json([
+            'username' => $user->getUsername(),
+            'displayName' => $user->getDisplayName(),
+            'roles' => $user->getRoles(),
+            'email' => $user->getEmail(),
+            'circles' => $user->getCircles(),
+        ]);
```

Comment:
This conversion to json is calling many functions on $user, a flat controller should only be the glue. Can the array of information for the ->json call be generated in the $user object?

```suggestion
        return $this->json($user->toArray());
```

Reply:
If $user was refering to a class I would agree with you, but here it's an interface. So a method toArray might refers to a serialization method and so expose sensitive information in the api

Reply:
I centralized the User to array conversion : 3d0a7d37515d4683568aae7b957bc3477a93cda7 A full refactoring of controller should be done and mainly refers to an ApiService

Reply:
I understand, but then make this function part of the userService. The controller should not have this logic.

Reply:
Moved in the service 9cbf75b7315e4fec4fa3edfeea82918a61565542

Reply:
even better in the user entity fda1219a36c1bec8f610426fdb8efc358057a658